File: Desktop_java/src/main/java/cnc/data/Package.java
package cnc.data;

public class Package {

	public static short TYPE_ABSOLUTE = 0;
	public static short TYPE_RELATIVE = 1;

	public short type;
	public short x, y, z;
	public short checkSum;

	public Package() {}

	public Package(short type, short x, short y, short z) {
		this.type = type;
		this.x = x;
		this.y = y;
		this.z = z;

		updateCheckSum();
	}

	public void updateCheckSum() {
		checkSum = type;
		checkSum ^= x << 1;
		checkSum ^= y << 2;
		checkSum ^= z << 3;
	}

}

File: MCU_arduino/DC_CNC/package.cpp

#include "cnc.h"

#define TYPE_ABSOLUTE 0
#define TYPE_RELATIVE 1


void executePackage(package* p){

  #ifdef DEBUG
  Serial.print("Execute: ");
  printPackage(p);
  #endif

  if(p->type == TYPE_ABSOLUTE){
    setupMoveXBy(p->x - state.x);
    setupMoveYBy(p->y - state.y);
    move();
  }
  else if(p->type == TYPE_RELATIVE){
    setupMoveXBy(p->x);
    setupMoveYBy(p->y);
    move();
  }
  else{ 
    Serial.println("Package type not defined!");
  }
  
}

void printPackage(package* p) {
  Serial.print("{type: ");
  Serial.print(p->type);
  Serial.print(", x: ");
  Serial.print(p->x);
  Serial.print(", y: ");
  Serial.print(p->y);
  Serial.print(", z: ");
  Serial.print(p->z);
  Serial.print(", sum: ");
  Serial.print(p->check_sum);
  Serial.println("};");
}

bool checkPackage(package* p){
  return check_sum(p) == p->check_sum;
}

int check_sum(package* p){
  int tmp = p->type;
  tmp ^= p->x << 1;
  tmp ^= p->y << 2;
  tmp ^= p->z << 3;
  return tmp;
}













